Add RabbitMQ Quorum queue migration playbook

Alex-Welsh opened the pull request:
This playbook is required for the upgrade job in Caracal CI

Tested & working

Alex-Welsh committed:
Add RabbitMQ Quorum queue migration playbook

Alex-Welsh committed:
Explicitly set RabbitMQ queue types in AIO

MoteHue commented on etc/kayobe/environments/ci-aio/kolla/globals.yml:
```diff
@@ -14,3 +14,7 @@ opensearch_heap_size: 200m
+
+# Ensure Rabbit is deployed with HA rather than Quorum queues (to test migrations)
+om_enable_rabbitmq_high_availability: true
+om_enable_rabbitmq_quorum_queues: false
```
This should be the default in stackhpc/2023.1?

Alex-Welsh replied:
Yes, but I want to set it explicitly so I can do [this](https://github.com/stackhpc/stackhpc-kayobe-config/pull/1203/files#diff-1102ab2bf041f53bc9ea3960af85f9852b3a4632c2e2d8bfa3bec097dc7090f8R317-R319)

Alex-Welsh committed:
Fix rabbit migration playbook

Alex-Welsh commented:
Tested & working

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+---
+# Migrate RabbitMQ queues from HA to Quorum
+# This is primarily used in CI workflows
```
Is this likely to be used outside of CI? If we don't want people to use it outside of CI we should mark it as such. If we do, perhaps it should be "the way"?

Alex-Welsh replied:
Refactored

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+    - name: Ensure that no queues exist
+      shell:
+        cmd: >
+          docker exec rabbitmq rabbitmqctl list_queues name --silent &&
```
If you want to grep the output of both commands you'll need to use a subshell:

(command1 && command2) | grep foo

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+    - name: Wait for the rabbitmq node to automatically start on container start
+      command: "docker exec {{ rabbitmq_container_name }} /bin/bash -c 'rabbitmqctl wait /var/lib/rabbitmq/mnesia/rabbitmq.pid --timeout 60'"
+      when: inspection.stdout == 'false'
+      become: true
```
This bit feels a bit weird. Why would rmq not be running at this point? Should we instead just check that it is running?

Alex-Welsh replied:
This just came from the rabbit reset playbook. It looked vaguely applicable so I slapped it in. I don't think it's of any benefit here so I've just removed it

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+      when: inspection.stdout == 'false'
+      become: true
+
+    - name: Generate new configuration and stop services
```
If this is likely to be run outside of CI it would be better to use a script to avoid nested ansible.

Alex-Welsh replied:
Scriptified

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+      shell:
+        cmd: "docker exec {{ rabbitmq_container_name }} rabbitmqctl list_queues type"
+      run_once: true
+      delegate_to: localhost
```
This won't work if rabbitmq is not running on localhost (which is the case in CI)

Alex-Welsh replied:
Refactored

markgoddard commented on etc/kayobe/ansible/migrate-rabbitmq-queues.yml:
```diff
@@ -0,0 +1,114 @@
+# This is primarily used in CI workflows
+
+- name: Migrate RabbitMQ queues
+  hosts: localhost
```
This isn't going to work in CI, because the github runner is the ansible control host, not the aio.

Alex-Welsh replied:
Refactored

markgoddard reviewed (changes requested):
I think overall I'd prefer to see this structured as 2 playbooks - prechecks and postchecks, and a script. If it's going to be explicitly marked as only for CI then I'd accept it in its current form.

Alex-Welsh commented:
> I think overall I'd prefer to see this structured as 2 playbooks - prechecks and postchecks, and a script. If it's going to be explicitly marked as only for CI then I'd accept it in its current form.

I'll restructure it

Alex-Welsh committed:
Refactor RabbitMQ migration playbook into script

markgoddard commented on etc/kayobe/ansible/rabbitmq-migration-postchecks.yml:
```diff
@@ -0,0 +1,25 @@
+  gather_facts: no
+  vars:
+    - rabbitmq_container_name: "rabbitmq"
+    - services_to_restart: "barbican,blazar,cinder,cloudkitty,designate,heat,ironic,keystone,magnum,manila,neutron,nova,octavia"
```
unused

Alex-Welsh replied:
Refactored

markgoddard commented on etc/kayobe/ansible/rabbitmq-migration-prechecks.yml:
```diff
@@ -0,0 +1,68 @@
+    - name: Generate kolla configuration
+      shell:
+        cmd: >
+          kayobe overcloud service configuration generate --node-config-dir /tmp/rabbit-migration --kolla-tags none
```
Move to script

Alex-Welsh replied:
Done

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,28 @@
+kayobe kolla ansible run "stop --yes-i-really-really-mean-it" -kt $RABBITMQ_SERVICES_TO_RESTART &&
+kayobe kolla ansible run rabbitmq-reset-state
+if [[ ! "$1" = "--skip-checks" ]]; then
+    kayobe overcloud host command run -l controllers -b --command "(docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_queues name --silent && docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_exchanges name --silent) | grep -v '^$' | (! grep -v 'amq.')"
```
Could be another playbook?

Alex-Welsh replied:
I've gone 100% scripty now to keep things cleaner

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,28 @@
+if [[ ! "$1" = "--skip-checks" ]]; then
+    kayobe overcloud host command run -l controllers -b --command "(docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_queues name --silent && docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_exchanges name --silent) | grep -v '^$' | (! grep -v 'amq.')"
+fi
+kayobe kolla ansible run deploy -kt {{ services_to_restart }}
```
`deploy-containers` would be quicker

Alex-Welsh replied:
done

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,28 @@
+#! /usr/bin/bash
+
```
`set -euo pipefail`

Alex-Welsh replied:
I've used -ex until now, testing pipefail and will see how it turns out on Monday morning

Alex-Welsh replied:
Seems fine

markgoddard replied:
No pipefail?

Alex-Welsh replied:
Reverted my changes :facepalm:
I'll quickly retest just to be sure

Alex-Welsh replied:
Good thing I checked. Setting pipefail permenantly kills the AIO somehow. It'll shut it down and then fail to start up again

Alex-Welsh replied:
Going to stick with -ex, I think that's the best we can do.

Alex-Welsh commented:
@markgoddard unfortunate timing on that review I'm afraid. I've just refactored it all into a script. It would've had to be at least 3 playbooks to cover all the checks I think, so I moved them all into the script and just use `host command run` to check controllers.

I was pushing to this branch just to make it easier to pull the changes to my AIOs for testing

Alex-Welsh commented:
General thoughts on the new script approach would be welcome though

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,56 @@
+kayobe kolla ansible run "stop --yes-i-really-really-mean-it" -kt $RABBITMQ_SERVICES_TO_RESTART &&
+kayobe kolla ansible run rabbitmq-reset-state
+if [[ ! "$1" = "--skip-checks" ]]; then
+    # Assert that Queues and exchanges have stopped (excluding those with 'amq.' in the name)
```
We might want a short sleep (10-30s?) here to allow any running services to register queues.

Alex-Welsh replied:
Done

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,56 @@
+kayobe kolla ansible run rabbitmq-reset-state
+if [[ ! "$1" = "--skip-checks" ]]; then
+    # Assert that Queues and exchanges have stopped (excluding those with 'amq.' in the name)
+    if kayobe overcloud host command run -l controllers -b --command \
+    "(docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_queues name --silent && \
+      docker exec $RABBITMQ_CONTAINER_NAME rabbitmqctl list_exchanges name --silent) | \
+      grep -v '^$' | (! grep -v 'amq.')"; then
+        echo "Failed check: RabbitMQ has not stopped properly"
+        exit 1
```
Would it be clearer to make this two separate commands, with the greps performed on the remote hosts?

Alex-Welsh replied:
Done

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,56 @@
+fi
+
+# Redeploy with Quorum Queues enabled
+kayobe kolla ansible run deploy -kt $RABBITMQ_SERVICES_TO_RESTART
```
We already generated config, so can use `deploy-containers`?

Alex-Welsh replied:
Done

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,56 @@
+# Redeploy with Quorum Queues enabled
+kayobe kolla ansible run deploy -kt $RABBITMQ_SERVICES_TO_RESTART
+
+if [[ ! "$1" = "--skip-checks" ]]; then
```
Another short sleep?

Alex-Welsh replied:
Done

Alex-Welsh committed:
Merge branch 'stackhpc/2023.1' into rabbit-upgrade

Alex-Welsh commented:
Again, tested and working

markgoddard commented on tools/rabbitmq-quorum-migration.sh:
```diff
@@ -0,0 +1,60 @@
+fi
+
+# Generate new config, stop services using rabbit, and reset rabbit state
+kayobe overcloud service configuration generate --node-config-dir /etc/kolla --kolla-skip-tags rabbitmq-ha-precheck &&
+kayobe kolla ansible run "stop --yes-i-really-really-mean-it" -kt $RABBITMQ_SERVICES_TO_RESTART &&
```
nit: don't need `&&` if you use `set -e`

Alex-Welsh replied:
Done

markgoddard reviewed (commented):
Mostly LGTM. Docs will need an update, but could be done in a separate PR.

MoteHue reviewed (dismissed):
Not tested myself, but looks good from my eyes

Alex-Welsh committed:
Address RabbitMQ quorum PR nits
